src/ir_pattern.py

```diff
@@ -0,0 +1,446 @@
+
+            next_try_matches = []
+            for state in try_matches:
+                # By pattern construction, pat_ref should be in the state's ref_map
+                # This would be true for "disjoint" or irrelevant instructions in the
+                # pattern, like random nops.
+                cand = state.try_map_ref(pat_ref)
+                if not isinstance(cand, InstrRef):
+                    continue
+                if not state.match_instr(pat_ref.instruction, cand.instruction):
+                    continue
+                states = state.permute_and_match_inputrefs(
+                    pat_inputs,
+                    flow_graph.instr_inputs[cand],
+                )
+                next_try_matches.extend(states)
+            try_matches = next_try_matches
+
+        for n, state in enumerate(try_matches):
+            # Perform any additional pattern-specific validation or computation
+            if not pattern.check(state):
+                continue
+
+            # Determine which instructions we can replace with the replacement_instr or nops
+            refs_to_replace: List[InstrRef] = [state.map_ref(tail_ref)]
+            for pat_ref in body_refs[::-1]:
+                cand_ref = state.map_ref(pat_ref)
+                # The candidate instruction can be replaced if all of its downstream
+                # uses are also being replaced
+                if all(
+                    all(r in refs_to_replace for r in refs)
+                    for _, refs in flow_graph.instr_uses[cand_ref].items()
+                ):
+                    refs_to_replace.append(cand_ref)
+
+            # Create temporary registers for the inputs to the replacement_instr
+            for pat_ref in input_refs:
+                assert len(pat_ref.instruction.outputs) == 1
+                input_reg = pat_ref.instruction.outputs[0]
+                assert isinstance(input_reg, Register)
+
+                original_reg = state.map_reg(input_reg)
+                temp_reg = Register(
+                    f"{original_reg.register_name}_fictive_{replace_index}"
+                )
+                state.rename_reg(input_reg, temp_reg)
+                move_instr = arch.parse(
+                    "move.fictive", [temp_reg, original_reg], InstructionMeta.missing()
+                )
+                input_uses = pattern.flow_graph.instr_uses[pat_ref].get(input_reg)
+                assert len(input_uses) >= 1
+                for use in input_uses:
+                    state.map_ref(use).add_instruction_before(move_instr)
+
+            for i, ref in enumerate(refs_to_replace):
+                # Remove ref from all instr_uses
+                instr = ref.instruction
+                for loc in instr.inputs:
+                    for r in flow_graph.instr_inputs[ref].get(loc):
+                        if isinstance(r, InstrRef):
+                            flow_graph.instr_uses[r].remove_ref(ref)
```

Comment:
and clear out instr_inputs[ref] to maintain graph correctness, and add input edges for the final instruction pointing to the in.fictive's

Reply:
can also assert that instr_uses[ref] is empty, except for i == 0. For i == 0 I believe instr_uses is kept? (but I'm not 100% sure. I don't remember seeing asserts for outputs being the same)

Reply:
instr_uses seems super easy to screw up when dynamically updating the graph. I wonder if we can add some error checking to it. Some ideas:
- recompute it from scratch and check that didn't change. Can be done until we switch translate.py over to the graph based info and make that the canonical source of truth. (heavy-weight option, annoying, slow)
- check that instr_inputs/instr_uses match up
- check that instr_inputs/instr_uses don't point to removed nodes/nops
- check that instr_inputs is non-empty if we have input registers (if that's even true)

Reply:
(I wish github allowed multiple review comment threads to hang off the same source line...)

Reply:
I refactored all the `instr_uses`/`instr_inputs` manipulation in a way that I think makes it easier to verify that it's correct, and I added the 2nd option you listed.

Now, whenever `instr_uses` is modified, `instr_uses` is also updated shortly afterwards. 

The assertion at the end seems to work: if I comment out some of the instr_uses/instr_inputs update statements, it trips the assert.